java/memory/src/test/java/org/apache/arrow/memory/TestBaseAllocator.java

```diff
@@ -222,7 +222,122 @@ public void testRootAllocator_createChildDontClose() throws Exception {
     }
   }
 
-  private static void allocateAndFree(final BufferAllocator allocator) {
+  // Allocation listener
+  // It counts the number of times it has been invoked, and how much memory allocation it has seen
+  // When set to 'expand on fail', it attempts to expand the associated allocator's limit
+  static class TestAllocationListener implements AllocationListener {
+    private int numCalls;
+    private long totalMem;
+    private boolean expandOnFail;
+    BufferAllocator expandAlloc;
+    long expandLimit;
+
+    TestAllocationListener() {
+      this.numCalls = 0;
+      this.totalMem = 0;
+      this.expandOnFail = false;
+      this.expandAlloc = null;
+      this.expandLimit = 0;
+    }
+
+    @Override
+    public void onAllocation(long size) {
+      numCalls++;
+      totalMem += size;
+    }
+
+    @Override
+    public boolean onFailedAllocation(long size,  Accountant.AllocationOutcome outcome) {
+      if (expandOnFail) {
+        expandAlloc.setLimit(expandLimit);
+        return true;
+      }
+      return false;
+    }
+
+    void setExpandOnFail(BufferAllocator expandAlloc, long expandLimit) {
+      this.expandOnFail = true;
+      this.expandAlloc = expandAlloc;
+      this.expandLimit = expandLimit;
+    }
+
+    int getNumCalls() {
+      return numCalls;
```

Comment:
you should be able to do this count number of calls monitoring using Jmockit's APIs rather than custom building something.

Reply:
True, and in fact I had it like that initially. However, I found I was repeating code again and again in the mocks, and it was much easier to just build a custom listener.

Reply:
ack